zerver/lib/markdown/__init__.py

```diff
@@ -572,6 +572,36 @@ def run(self, root: Element) -> None:
             video.set("data-video-original-url", url)
 
 
+class InlineAudioProcessor(markdown.treeprocessors.Treeprocessor):
+    """
+    Rewrite inline audio tags to serve external content via Camo.
+
+    This rewrites all audio, except ones that are served from the current
+    realm or global STATIC_URL. This is to ensure that each realm only loads
+    audios that are hosted on that realm or by the global installation,
+    avoiding information leakage to external domains or between realms. We need
+    to disable proxying of audios hosted on the same realm, because otherwise
+    we will break audios in /user_uploads/, which require authorization to
+    view.
+    """
+
+    def __init__(self, zmd: "ZulipMarkdown") -> None:
+        super().__init__(zmd)
+        self.zmd = zmd
+
+    @override
+    def run(self, root: Element) -> None:
+        # Get all URLs from the blob
+        found_audios = walk_tree(root, lambda e: e if e.tag == "audio" else None)
+        for audio in found_audios:
+            url = audio.get("src")
+            assert url is not None
+            if is_static_or_current_realm_url(url, self.zmd.zulip_realm):
+                continue
+            audio.set("src", get_camo_url(url))
+            audio.set("data-audio-original-url", url)
```

Comment:
To what extent does this function differ from the similar one for images?

Reply:
@timabbott , this function is rather similar to the ```InlineVideoProcessor``` one, and I have picked it up, since both show nearly the same behaviour while looking up for elements in the Tree and setting the urls.
The images one just has ```camo url``` since there the support is for upload from device, while here we're dealing with upload from links as well, so the original audio url is also included.